Added MultilayerPerceptron

Abi1024 opened the pull request:
Multilayer Neural Network implementation for Saul, with support for sparse feature vectors, various activation functions and L-2 regularization.

kordjamshidi commented:
Thank you very much @Abi1024 . @bhargav @danyaljj please review at your earliest convenience.

kordjamshidi commented on saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java:
```diff
@@ -0,0 +1,453 @@
+    }
+
+    public ScoreSet scores(Object example){
+        return null;
```
We need scores here but this returns null, right?

danyaljj replied:
If I understand correctly we have to call the `infer` function here?

danyaljj commented on saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java:
```diff
@@ -0,0 +1,453 @@
+    }
+
+    //Construct a perceptron with user specified parameters.
+    public MultilayerPerceptron(int num_inputs, int num_outputs, int[] hidden_layers, int [] activation_function){
```
Could you clarify each input value with a javadoc description?

danyaljj commented on saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java:
```diff
@@ -0,0 +1,453 @@
+    private static int default_num_inputs = 128;
+    private static int[] default_hidden_layers =  new int[]{};
+    private static int[] default_activation_function = {SIGMOID};
+    private static int default_num_outputs = 2;
```
Does "2" here mean "one output variables for 2 classes"?

danyaljj commented on saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java:
```diff
@@ -0,0 +1,453 @@
+    }
+
+    public void write(ExceptionlessOutputStream out) {
+        out.close();
```
Can you add proper TODO message here, so that we remember to implement it? 
To be honest classifiers are not that useful in Saul, if their serialization/deserialization is not working. 

FYI @kordjamshidi

danyaljj commented on saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java:
```diff
@@ -0,0 +1,588 @@
+    /**
+     * Trains the neural network on a sparse input vector, using default learning parameters.
+     *
+     * @TODO Only works for binary classification now.
```
could you clarify this? Do you mean, the output vector can only be a vector of 1.0/0.0 ?

danyaljj commented:
Looks (almost) good to me. Passing it to @bhargav for a final look.

bhargav commented on saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java:
```diff
@@ -0,0 +1,583 @@
+
+        // Initializes a layer of the neural network. Weights are all initialized to 0.
+        private Layer(int num_inputs, int num_outputs, int activation_function){
+            weights = new double[num_inputs+1][num_outputs];
```
Minor: I believe primitives in Java are initialized to 0 by default. You might not need to fill the arrays with 0.

bhargav commented on saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java:
```diff
@@ -0,0 +1,583 @@
+            }
+        }
+        if (best_index == 0){ //for binary classifiers used in Twitter sentiment analysis
+            return new DiscretePrimitiveStringFeature(containingPackage, name, "",
```
The returned features here would be problematic for other tasks. The labels might not be "negative"/"positive"

bhargav replied:
You should use something to the effect of `labelLexicon.lookupKey(i)` on the best_index.

bhargav commented on saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java:
```diff
@@ -0,0 +1,583 @@
+            this.activation_function = activation_function;
+        }
+
+        public void write(ExceptionlessOutputStream out){
```
Quick Question: while serializing models, we should not save the gradients/prev_gradients and activations right? We will have to evaluate them on examples in the forward(-backwards) pass when we load a model to test or train.

bhargav commented on saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java:
```diff
@@ -0,0 +1,583 @@
+            }
+        }
+        if (layers[0].weights.length < maximum_index + 2 )
+            resize_inputs(maximum_index + 10000);
```
For later: It might be better to do a pre-extract step and calculate the final label lexicon size. This might add a high computation overhead if we do not have 10k more features.

Abi1024 commented:
Hi, I just incorporated all of @bhargav 's suggestions and also added regression for the facebook circumplex dataset. If someone could take a look sometime, that would be much appreciated. Thanks!

kordjamshidi commented:
@bhargav please see if this is ok with the recent changes. However, I am not sure if we want to merge the new app of circumplex (sentiment analysis) model also into Saul or not. @Abi1024 can remove that application, maybe, and just keep the implementation of the Learners. Let him know if removing the App is preferred @danyaljj @bhargav .

kordjamshidi commented:
Oh, I see @bhargav has approved the changes already. So we merge the app too? Any opinion, comments?

danyaljj commented:
Agreed. Unless the application is complete  (fully functional, with readme etc), I would rather not checking it in.

kordjamshidi commented:
Thanks @Abi1024 for removing the app, I m going to merge then.
